File: JS/3.longest-substring-without-repeating-characters.js
/*
 * @lc app=leetcode id=3 lang=javascript
 *
 * [3] Longest Substring Without Repeating Characters
 */

// @lc code=start
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
  let len = s.length;
  let map = new Map();
  let left = 0,
    right = 0;
  let maxLen = Number.MIN_VALUE;
  while (right < len) {
    if (map.has(s[right])) {
      left = Math.max(left, map.get(s[right]) + 1); // Go to the last appeared left
    }
    maxLen = Math.max(right - left + 1, maxLen);
    map.set(s[right], right);
    right++;
  }
  return maxLen;
};
// @lc code=end

File: Rust/3.longest-substring-without-repeating-characters.rs
/*
 * @lc app=leetcode id=3 lang=rust
 *
 * [3] Longest Substring Without Repeating Characters
 */

// @lc code=start
impl Solution {
    pub fn length_of_longest_substring(s: String) -> i32 {
        use std::collections::HashMap;
        use std::cmp::max;
        let mut map: HashMap<char, i32> = HashMap::with_capacity(s.len());
        let (mut left, mut right, mut maxLen) = (0, 0, 0);
        for c in s.chars() {
            if(map.contains_key(&c)) {
                if let Some(j) = map.get(&c) {
                    left = max(left, j+1);
                }
            }
            maxLen = max(right - left + 1, maxLen);
            map.insert(c, right);
            right+=1;
        } 
        
        maxLen
    }
}
// @lc code=end

